File: python/sqlalchemy/left_join_demo.py
"""LEFT JOIN demo for SQLAlchemy builder scanning."""

from sqlalchemy.orm import Session

from models import Order, User


def list_user_order_status(session: Session):
    """Return joined user/order rows with a bounded LEFT JOIN query."""
    return (
        session.query(User.id, User.email, Order.status)
        .outerjoin(Order, Order.user_id == User.id)
        .filter(User.active.is_(True))
        .limit(25)
        .all()
    )


def list_broken_user_order_status(session: Session):
    """Return a broken LEFT JOIN query that should trigger schema findings."""
    return (
        session.query(User.id, User.email, Order.ghost_status)
        .outerjoin(Order, Order.user_id == User.id)
        .filter(Order.missing_flag == "pending")
        .all()
    )

File: go/goqu/left_join_demo.go
// Package goqu demonstrates LEFT JOIN query patterns with Goqu.
package goqu

import (
	"context"
	"database/sql"

	goqulib "github.com/doug-martin/goqu/v9"
)

// JoinAuditStore demonstrates safe and broken LEFT JOIN queries.
type JoinAuditStore struct {
	db *goqulib.Database
}

// NewJoinAuditStore creates a new JoinAuditStore.
func NewJoinAuditStore(db *sql.DB) *JoinAuditStore {
	return &JoinAuditStore{db: goqulib.New("postgres", db)}
}

// ListUserOrderStatus demonstrates a safe LEFT JOIN query.
func (s *JoinAuditStore) ListUserOrderStatus(ctx context.Context) error {
	_ = ctx
	_, _, err := goqulib.From("users").
		LeftJoin(
			goqulib.T("orders"),
			goqulib.On(goqulib.I("orders.user_id").Eq(goqulib.I("users.id"))),
		).
		Select("users.id", "users.email", "orders.status").
		Where(goqulib.I("users.active").Eq(true)).
		Limit(25).
		ToSQL()
	return err
}

// ListBrokenUserOrderStatus demonstrates a broken LEFT JOIN query.
func (s *JoinAuditStore) ListBrokenUserOrderStatus(ctx context.Context) error {
	_ = ctx
	_, _, err := goqulib.From("users").
		LeftJoin(
			goqulib.T("orders"),
			goqulib.On(goqulib.I("orders.user_id").Eq(goqulib.I("users.id"))),
		).
		Select("users.id", "users.email", "orders.ghost_status").
		Where(goqulib.I("orders.missing_flag").Eq("pending")).
		ToSQL()
	return err
}
